src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt

```diff
@@ -220,50 +222,113 @@ class LicenceService(
+    val revisedConditions = licenceEntity.additionalConditions.filter { it.id != conditionId }
+
+    val updatedLicence = licenceEntity.copy(
+      additionalConditions = revisedConditions,
+      dateLastUpdated = LocalDateTime.now(),
+      updatedByUsername = username,
+    )
+
+    licenceRepository.saveAndFlush(updatedLicence)
+  }
+
   @Transactional
   fun updateAdditionalConditions(licenceId: Long, request: AdditionalConditionsRequest) {
     val licenceEntity = licenceRepository
       .findById(licenceId)
       .orElseThrow { EntityNotFoundException("$licenceId") }
 
     val username = SecurityContextHolder.getContext().authentication.name
-    val additionalConditions = licenceEntity.additionalConditions.associateBy { it.conditionCode }.toMutableMap()
-    val newAdditionalConditions = request.additionalConditions.transformToEntityAdditional(licenceEntity, request.conditionType)
-
-    // Update any existing additional conditions with new values, or add the new condition if it doesn't exist.
-    newAdditionalConditions.forEach {
-      if (additionalConditions[it.conditionCode] != null) {
-        additionalConditions[it.conditionCode]?.conditionCategory = it.conditionCategory
-        additionalConditions[it.conditionCode]?.conditionText = it.conditionText
-        additionalConditions[it.conditionCode]?.conditionSequence = it.conditionSequence
-        additionalConditions[it.conditionCode]?.conditionType = it.conditionType
-      } else {
-        it.licence = licenceEntity
-        additionalConditions[it.conditionCode] = it
-        additionalConditions[it.conditionCode]?.expandedConditionText = it.conditionText
-      }
-    }
 
-    // Remove any additional conditions which exist on the licence, but were not specified in the request
-    val resultAdditionalConditionsList = additionalConditions.values.filter { condition ->
-      newAdditionalConditions.find { newAdditionalCondition -> newAdditionalCondition.conditionCode == condition.conditionCode } != null ||
-        condition.conditionType != request.conditionType
+    val submittedAdditionalConditions =
+      request.additionalConditions.transformToEntityAdditional(licenceEntity, request.conditionType)
+
+    val newAdditionalConditions =
+      submittedAdditionalConditions.filter { licenceEntity.additionalConditions.none { submittedCondition -> submittedCondition.conditionCode == it.conditionCode } }
+        .map { newAdditionalCondition -> newAdditionalCondition.copy(expandedConditionText = newAdditionalCondition.conditionText, licence = licenceEntity) }
+
+    val removedAdditionalConditionsList =
+      licenceEntity.additionalConditions.filter { submittedAdditionalConditions.none { submittedCondition -> submittedCondition.conditionCode == it.conditionCode } }
+
+    val updatedAdditionalConditions =
+      licenceEntity.additionalConditions.filter { condition -> removedAdditionalConditionsList.none { it.conditionCode == condition.conditionCode } }
```

Comment:
mmm.. could possibly add an `doesNotExistIn` helper extension function? 

```
licenceEntity.additionalConditions.filter { it.doesNotExistIn(removedAdditionalConditionsList) }
```

I think you'd be able to use it 3 times.. maybe the double negation in a filter is a bit brain blowing! 

Not super essential, just a bit difficult to read.